design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+synchronization as however they'd like.
+
+> Though the override behavior also paves the way to target a completely
+> different central cluster, too, which may break the setup since CRDs in that
+> other central cluster could be different. We accept this risk as it would have
+> to be a deliberate choice of the user.
+
+#### Authorization
+
+##### Remote Cluster
+
+Since there will be one agent for the whole cluster, its own mounted
+`ServiceAccount` in that remote cluster needs to have read & write permissions
+for all of the following kinds in the remote cluster listed below:
+
+* `CustomResourceDefinition`
+* `Composition`
+* `InfrastructureDefinition`
+* `InfrastructurePublication`
+* `Secret`
+* All `*Requirement` types
+
+The last one is a bit tricky because the exact list of `*Requirement` types on
+kind level is not known during installation and it's not static; new published
+APIs should be available in the remote cluster dynamically. One option is to
+allow agent to grant `Role` and `RoleBinding`s to itself as it creates the
+necessary `CustomResourceDefinition`s in the remote cluster. However, an entity
+that is able to grant permissions to itself could greatly increase the security
+posture.
+
+When you zoom out and think about how the `Role` will look like, in most of the
+cases, it's something like the following:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: crossplane-agent
+  namespace: default
+rules:
+  # all kinds could be under one company/organization group
+- apiGroups: ["acmeco.org"] 
+  resources: ["*"]
+  verbs: ["*"]
+  # or there could be logical groupings for different sets of requirements
+- apiGroups: ["database.acmeco.org"]
+  resources: ["*"]
+  verbs: ["*"]
+```
+
+As you can see, it's either one group for all the new APIs or a logical group
+for each set of APIs. In both cases, the frequency of the need to add a new
+`apiGroup` is less than one would imagine thanks to the ability of allowing a
+whole group; most frequently, the platform operators will be adding new kinds to
+the existing groups.
+
+In the light of this assumption, the initial approach will be that the `Role`
+bound to the agent will be populated by a static list of the groups of the
+requirement types during the installation like shown above and if a new group is
+introduced, then an addition to this `Role` will be needed. A separate
+controller to dynamically manage the `Role` is mentioned in the [Future
+Considerations](#future-considerations) section.
+
+##### Central Cluster
+
+The `ServiceAccount` that will be created in the central cluster needs to have
+the following permissions for agent to do its operations:
+
+* Read
+  * `CustomResourceDefinition`s
+  * `InfrastructureDefinition`s
+  * `InfrastructurePublication`s
+  * `Composition`s
+  * `Secret`s in the given namespace.
+* Write
+  * `*Requirement` types that you'd like to allow in given namespace.
```

Comment:
> To be clear this seems like n namespaced RoleBindings to a Roles for a given requirement, and might be hard in practice to calculate, so if the agent is adding annotations dynamically.

I think you can have one `Role` in the target namespace of the central cluster that includes all of your `*Requirement` types. In the example below, we add all `kind`s that exist under `database.acmeco.org` and `acmeco.org` groups. I couldn't get the annotation part though; agent doesn't annotate any resource on its own.

> I think something that declares the usage of a requirement might be helpful here to simplify the model.

Could you elaborate that with an example?

Reply:
As an app-operator on an agent cluster it would be nice to clearly know what requirements I can operate on. In my experience with CRDs and RBAC this is not a user-friendly interface per se. I'll leave it as a conceptual idea whether the agent should consider a higher level concept of a list of requirements that are allowed.

Reply:
Got it. The UX of CRDs and RBAC could be improved with higher level abstractions but I'd tackle that problem with another one-pager/issue instead where we can consider many options and think carefully about not complicating the system.